models/demos/t3000/falcon40b/tests/test_perplexity_falcon.py

```diff
@@ -0,0 +1,303 @@
+    configuration = hugging_face_reference_model.config
+
+    # Prepare dataset
+    logger.info("Preparing dataset...")
+    dataset = prepare_textgen_dataset(dataset_name, dataset_config, split)
+    tokenizer = AutoTokenizer.from_pretrained(model_version)
+    encodings = tokenizer(dataset, return_tensors="pt")["input_ids"].squeeze(0)
+    dataloader = prepare_textgen_dataloader(encodings, batch_size, max_seq_len, num_samples, stride)
+
+    if not use_hf_model:
+        # Load tt-metal model config
+        input_shape = [batch_size, max_seq_len]
+        model_config = get_model_config(
+            model_config_str, llm_mode, input_shape, num_devices=len(device_mesh.get_devices())
+        )
+        tt_cache_path = get_tt_cache_path(
+            model_version, model_subdir="Falcon", default_dir=model_config["DEFAULT_CACHE_PATH"]
+        )
+
+        # Load tt-metal model
+        logger.info("Moving weights (all layers) to device; might take some time...")
+        model = TtFalconCausalLM(
+            device_mesh,
+            state_dict,
+            "",
+            num_layers,
+            configuration,
+            max_seq_len,
+            model_config,
+            tt_cache_path,
+            use_global_cos_sin_cache=True,
+        )
+        for device in device_mesh.get_devices():
+            ttnn.device.synchronize_device(device)
+
+        # Initialize kvcache
+        logger.info("Initializing kvcache...")
+        kv_cache = model.initialize_kv_cache()
+    else:
+        # model = pytorch_FalconCausalLM
+        model = hugging_face_reference_model
+        kv_cache = None
+
+    # Evaluate perplexity
+    logger.info("Evaluating perplexity...")
+    start = time.time()
+    nll, ppl, top1_acc, top5_acc = calculate_perplexity(
+        model,
+        dataloader,
+        llm_mode,
+        batch_size,
+        max_seq_len,
+        kv_cache,
+        configuration,
+        device_mesh,
+        use_hf_model=use_hf_model,
+    )
+    logger.info(f"Perplexity evaluation time: {(time.time() - start):.2f} s")
+    logger.info(f"Negative log-likelihood: {nll:.4f}")
+    logger.info(f"Perplexity: {ppl:.4f}")
+    logger.info(f"Top-1 accuracy: {top1_acc:.4f}")
+    logger.info(f"Top-5 accuracy: {top5_acc:.4f}")
+
+    # Verify metrics against targets
+    calculated_acc_metrics = {"ppl": ppl, "top1_acc": top1_acc, "top5_acc": top5_acc}
+    verify_acc_metrics(calculated_acc_metrics, expected_acc_metrics)
+
+
+@pytest.mark.parametrize(
+    "llm_mode, batch_size, max_seq_len, num_samples, expected_ppl, expected_top1, expected_top5",
+    (
+        ("prefill", 32, 128, 64, 12.67, 0.47, 0.71),
+        ("prefill", 32, 1024, 64, 7.21, 0.55, 0.79),
+        ("prefill", 32, 2048, 64, 9.81, 0.50, 0.74),  # TODO: run for falcon40b
+        ("decode", 64, 128, 64, 12.67, 0.47, 0.71),
+        ("decode", 64, 1024, 64, 7.21, 0.55, 0.79),
+        ("decode", 64, 2048, 64, 9.81, 0.50, 0.74),  # TODO: run for falcon40b
```

Comment:
What do these todos refer to?

Reply:
I'm currently running the falcon40b reference on CPU, it takes ~2.5 days so I'll update the numbers as soon as I have them - those tests aren't run in CI anyway. The numbers you see here are from falcon7b.